base-station/ir_interface.py

```diff
@@ -47,15 +45,377 @@
+        self.baudrate = config.get(key="usb_baudrate")
+        self.timeout = config.get(key="usb_timeout")
+        self.failure_try = config.get(key="usb_failure_try")
+        self.failure_wait = config.get(key="usb_failure_wait")
+        self.packet_que_max = config.get(key="packet_que_max")
+        self.duplex = config.get(key="duplex")
+        self.seq_set = set()
+        self.current_seq = 0
+        self.pt_seq_que_table = dict([(e, dict()) for e in (PT.QTR,)] + 
+            [(e, asyncio.Queue(maxsize=self.packet_que_max)) for e in (PT.RRR, PT.GSR)])
+        self.write_lock = asyncio.Lock()
+        self.read_lock = asyncio.Lock()
+        self.half_duplex_lock = asyncio.Lock()
+        self.badge_status: Packet = None
+
+        try:
+            self.serial = serial.Serial(
+                port=self.port,
+                baudrate=self.baudrate,
+                timeout=self.timeout,
+                write_timeout=self.timeout
+            )
+        except serial.SerialException as e:
+            print(f"Failed to initialize serial port: {e}")
+            raise
 
-    async def trigger_send_packet(self, data: bytes) -> bool:
+    async def trigger_send_packet(self, data: bytes, packet_type = PT.QTQ) -> bool | Packet:
         # Triggers the ir interface to send a packet.
         # Returns True if sent successfully, False otherwise.
-        pass
+        for _ in range(self.failure_try):          
+            if not self.serial.is_open:
+                self.serial.open()
+                if not self.serial.is_open:
+                    await asyncio.sleep(self.failure_wait)
+                    continue
+            
+            try:
+                result: bool | Packet = await self._write_packet(data, packet_type)
+                if type(result) == Packet and result.is_success == IrInterface.success:
+                    return result
+            
+                if result == True:
+                    return True
+            except Exception as e:
+                print(f"Error sending packet: {e}")
+            
+            await asyncio.sleep(self.failure_wait)
+
+        return False
+    
 
     async def get_next_packet(self) -> bytes:
-        # Wait until the next packet arrives, then return its raw data.
-        pass
+        # Wait until the next packet arrives, then return its raw data.    
+        for _ in range(self.failure_try):
+            if not self.serial.is_open:
+                self.serial.open()
+                if not self.serial.is_open:
+                    await asyncio.sleep(self.failure_wait)
+                    continue
+
+            try:
+                return await self._read_packet() # May idle
+            except Exception as e:
+                print(f"Error getting packet: {e}")
+                await asyncio.sleep(self.failure_wait)
+
+        return b''
+    
+
+    async def _write_packet(self, data: bytes, packet_type = PT.QTQ, lock = True) -> bool | Packet:
```

Comment:
Optional: Consider splitting this into _write_packet() and _write_packet_and_wait_for_response()